src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs

```diff
@@ -0,0 +1,301 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+using Microsoft.AspNet.OData.Common;
+using Microsoft.AspNet.OData.Formatter;
+using Microsoft.AspNet.OData.Formatter.Serialization;
+using Microsoft.AspNet.OData.Query.Expressions;
+using Microsoft.OData;
+using Microsoft.OData.Edm;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.AspNet.OData.Query
+{
+    /// <summary>
+    /// Default implementation of SkipTokenHandler for the service. 
+    /// </summary>
+    public class DefaultSkipTokenHandler : SkipTokenHandler
+    {
+        private IDictionary<string, object> _propertyValuePairs;
```

Comment:
I have a concern about storing the values of the current request in the skip token handler. This, in conjunction with the scope of the skip token handler being Singleton, will prevent concurrent requests to work properly, because concurrent requests will write in the same dictionary.

My suggestion is to use the return value of the method `ProcessSkipTokenValue` (which is currently dismissed by the caller), and pass it to the `Apply` methods.

Reply:
Sorry don't mind the comment I mixed up the pull requests.